substrate/frame/assets-holder/src/impls.rs

```diff
@@ -0,0 +1,253 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+use frame_support::traits::{
+	fungibles::{Dust, Inspect, InspectHold, MutateHold, Unbalanced, UnbalancedHold},
+	tokens::{
+		DepositConsequence, Fortitude, Precision, Preservation, Provenance, WithdrawConsequence,
+	},
+};
+use pallet_assets::HeldBalance;
+use sp_runtime::{
+	traits::{CheckedAdd, CheckedSub, Zero},
+	ArithmeticError,
+};
+
+// Implements [`HeldBalance`] from [`pallet-assets`], so it can understand whether there's a held
+// amount for an asset account, and is able to signal to this pallet when to clear the state of an
+// account.
+impl<T: Config<I>, I: 'static> HeldBalance<T::AssetId, T::AccountId, T::Balance> for Pallet<T, I> {
+	fn held_balance(asset: T::AssetId, who: &T::AccountId) -> Option<T::Balance> {
+		HeldBalances::<T, I>::get(asset, who)
+	}
+
+	fn died(asset: T::AssetId, who: &T::AccountId) {
+		HeldBalances::<T, I>::remove(asset.clone(), who);
+		Holds::<T, I>::remove(asset, who);
+	}
+}
+
+// Implement [`fungibles::Inspect`](frame_support::traits::fungibles::Inspect) as it is bound by
+// [`fungibles::InspectHold`](frame_support::traits::fungibles::InspectHold) and
+// [`fungibles::MutateHold`](frame_support::traits::fungibles::MutateHold). To do so, we'll
+// re-export all of `pallet-assets` implementation of the same trait.
+impl<T: Config<I>, I: 'static> Inspect<T::AccountId> for Pallet<T, I> {
+	type AssetId = T::AssetId;
+	type Balance = T::Balance;
+
+	fn total_issuance(asset: Self::AssetId) -> Self::Balance {
+		pallet_assets::Pallet::<T, I>::total_issuance(asset)
+	}
+
+	fn minimum_balance(asset: Self::AssetId) -> Self::Balance {
+		pallet_assets::Pallet::<T, I>::minimum_balance(asset)
+	}
+
+	fn total_balance(asset: Self::AssetId, who: &T::AccountId) -> Self::Balance {
+		pallet_assets::Pallet::<T, I>::total_balance(asset, who)
```

Comment:
in pallet assets this is implemented by calling the `balance` function.

Reply:
This is the equivalent method in `pallet-balances`. See lines `57`, `69`, and `72`.

https://github.com/paritytech/polkadot-sdk/blob/711d91aaabb9687cd26904443c23706cd4dbd1ee/substrate/frame/balances/src/impl_fungible.rs#L47-L73

Reply:
Yes! That's a different function from `reducible_balance`. While `balance` is used to inspect the total balance of an asset account, `reducible_balance` is used to determine whether a `balance` can be decreased by an `amount`.

Reply:
Change applied!